### ISSUE TITLE
Using Vis with GWT

### ISSUE THREAD

**Issue by zbeedatm** [almende/vis#3617]
Hi,
I'm trying to use vis from my gwt application. I'll be calling an api that will retrieve my tree data in some structure, and I thought to prepare data parameters for vis.js method and call it within JSNI. PSB:

`import com.google.gwt.core.client.EntryPoint;
import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONParser;
import com.google.gwt.json.client.JSONValue;
import com.google.gwt.visualization.client.DataTable;
import com.google.gwt.visualization.client.DataView;

/*String[] arrNodes = {
				"{id: 1, label: 'Node 1'}",
				"{id: 2, label: 'Node 2'}",
				"{id: 3, label: 'Node 3'}",
				"{id: 4, label: 'Node 4'}",
				"{id: 5, label: 'Node 5'}"
		};
		
		String[] arrEdges = {
				"{from: 1, to: 3}",
				"{from: 1, to: 2}",
				"{from: 2, to: 4}",
				"{from: 2, to: 5}",
				"{from: 3, to: 3}"
		};*/

com.google.gwt.json.client.JSONObject jsonObj = new com.google.gwt.json.client.JSONObject();
		JSONValue jsonValue = JSONParser.parseStrict("1");
		jsonObj.put("from",jsonValue);
		jsonValue = JSONParser.parseStrict("2");
		jsonObj.put("to",jsonValue);
		
		DataTable dataTable = DataTable.create(jsonObj.getJavaScriptObject());
		DataView dvEdges = DataView.create(dataTable);
		
		jsonObj = new JSONObject();
		jsonValue = JSONParser.parseStrict("1");
		jsonObj.put("id",jsonValue);
		jsonValue = JSONParser.parseStrict("Node1");
		jsonObj.put("label",jsonValue);
		jsonValue = JSONParser.parseStrict("2");
		jsonObj.put("id",jsonValue);
		jsonValue = JSONParser.parseStrict("Node2");
		jsonObj.put("label",jsonValue);
		
		dataTable = DataTable.create(jsonObj.getJavaScriptObject());
		DataView dvNodes = DataView.create(dataTable);
		
		
		loadNetwork(dvNodes, dvEdges);
	}
	
	
	private native void loadNetwork(DataView nodes, DataView edges)/*-{  
	
	  // create a network
	  var container = $doc.getElementById('mynetwork');
	  var data = {
	    nodes: nodes,
	    edges: edges
	  };
	  var options = {};
	  var network = new $wnd.vis.Network(container, data, options);
	  
		$wnd.draw();
	}-*/;
`
When I passed the parameters as string, the vis script was not able to read it. It's expecting a DataSet/DataView.
In order to work with DataTable/DataView in Java I tried using the gwt-visualation jar, but still getting errors:

> DataTable.java:43 Uncaught TypeError: Cannot read property 'visualization' of undefined
    at create_61_g$ (DataTable.java:43)
    at GWT_VisJs_Manuplation_Demo_1_g$.onModuleLoad_3_g$ [as onModuleLoad_4_g$] (GWT_VisJs_Manuplation_Demo.java:131)
    at Array.init_0_g$ (com_00046gwt_00046vis_00046GWT__VisJs__Manuplation__Demo__EntryMethodHolder.java:3)
    at initializeModules_0_g$ (ModuleUtils.java:44)
    at apply_0_g$ (Impl.java:247)
    at entry0_0_g$ (Impl.java:306)
    at Impl.java:72
    at gwtOnLoad_0_g$ (ModuleUtils.java:55)
    at ModuleUtils.java:25
create_61_g$ @ DataTable.java:43
onModuleLoad_3_g$ @ GWT_VisJs_Manuplation_Demo.java:131
init_0_g$ @ com_00046gwt_00046vis_00046GWT__VisJs__Manuplation__Demo__EntryMethodHolder.java:3
initializeModules_0_g$ @ ModuleUtils.java:44
apply_0_g$ @ Impl.java:247
entry0_0_g$ @ Impl.java:306
(anonymous) @ Impl.java:72
gwtOnLoad_0_g$ @ ModuleUtils.java:55
(anonymous) @ ModuleUtils.java:25

Can you please assist...

B.R.

**Comment by zbeedatm** [comment]
I did the following changes #

to my host page html I added;
`<script type="text/javascript" src="https://www.google.com/jsapi"></script>
<script type="text/javascript">
  	google.load("visualization", "1", {packages:["corechart"]});`

and that's did the treck for the google.visualation issue, but still have a warning:

> A Parser-blocking, cross site (i.e. different eTLD+1) script, https://www.google.com/uds/?file=visualization&v=1&packages=corechart, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message.See https://www.chromestatus.com/feature/5718547946799104 for more details.

And I realized that I was building my JSONObject not in the correct way. I changed the code to:
`JSONArray arrNodes = new JSONArray();
		JSONValue jsonValue = JSONParser.parseStrict("{\"id\": 1, \"label\": \"Node 1\"}");
		GWT.log("JSON Value ====> " + jsonValue.toString());
		arrNodes.set(0, jsonValue);
		jsonValue = JSONParser.parseStrict("{\"id\": 2, \"label\": \"Node 2\"}");
		arrNodes.set(1, jsonValue);
		
		JSONArray arrEdges = new JSONArray();
		jsonValue = JSONParser.parseStrict("{\"from\": 1, \"to\": 2}");
		arrEdges.set(0, jsonValue);
		
		
		DataTable dataTable = DataTable.create(arrEdges.getJavaScriptObject());
		DataView dvNodes = DataView.create(dataTable);
		
		dataTable = DataTable.create(arrNodes.getJavaScriptObject());
		DataView dvEdges = DataView.create(dataTable);`


Now the error is in vis.js #3450 

> vis.js:44115 Uncaught TypeError: Array or DataSet expected
    at NodesHandler.setData (vis.js:44115)
    at Network.setData (vis.js:42781)
    at new Network (vis.js:42571)
    at GWT_VisJs_Manuplation_Demo_1_g$.loadNetwork_0_g$ [as loadNetwork_1_g$]


`    /**
     * Set a data set with nodes for the network
     * @param {Array | DataSet | DataView} nodes         The data containing the nodes.
     * @param {boolean} [doNotEmit=false]
     * @private
     */

  }, {
    key: 'setData',
    value: function setData(nodes) {
      var doNotEmit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

      var oldNodesData = this.body.data.nodes;

      if (nodes instanceof DataSet || nodes instanceof DataView) {
        this.body.data.nodes = nodes;
      } else if (Array.isArray(nodes)) {
        this.body.data.nodes = new DataSet();
        this.body.data.nodes.add(nodes);
      } else if (!nodes) {
        this.body.data.nodes = new DataSet();
      } else {
        throw new TypeError('Array or DataSet expected');
      }`

**Comment by zbeedatm** [comment]
I did the following:

Added to my host page html:
<script type="text/javascript" src="https://www.google.com/jsapi"></script>
script type="text/javascript">
google.load("visualization", "1", {packages:["corechart"]});

And that helped getting rid of the google.visualization load error, but still having a warning:

> A Parser-blocking, cross site (i.e. different eTLD+1) script, https://www.google.com/uds/?file=visualization&v=1&packages=corechart, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message.See https://www.chromestatus.com/feature/5718547946799104 for more details.

I realized as well that my code wasn't creating the JSONObject as expected, so it was changed to:
`JSONArray arrNodes = new JSONArray();
		JSONValue jsonValue = JSONParser.parseStrict("{\"id\": 1, \"label\": \"Node 1\"}");
		GWT.log("JSON Value ====> " + jsonValue.toString());
		arrNodes.set(0, jsonValue);
		jsonValue = JSONParser.parseStrict("{\"id\": 2, \"label\": \"Node 2\"}");
		arrNodes.set(1, jsonValue);
		
		JSONArray arrEdges = new JSONArray();
		jsonValue = JSONParser.parseStrict("{\"from\": 1, \"to\": 2}");
		arrEdges.set(0, jsonValue);
		
		
		DataTable dataTable = DataTable.create(arrEdges.getJavaScriptObject());
		DataView dvNodes = DataView.create(dataTable);
		
		dataTable = DataTable.create(arrNodes.getJavaScriptObject());
		DataView dvEdges = DataView.create(dataTable);`

But now I'm failing in vis.js code:

> Uncaught TypeError: Array or DataSet expected
    at NodesHandler.setData (vis.js:44115)
    at Network.setData (vis.js:42781)
    at new Network (vis.js:42571)
    at GWT_VisJs_Manuplation_Demo_1_g$.loadNetwork_0_g$ [as loadNetwork_1_g$] (GWT_VisJs_Manuplation_Demo.java:223)

The code:
`    /**
     * Set a data set with nodes for the network
     * @param {Array | DataSet | DataView} nodes         The data containing the nodes.
     * @param {boolean} [doNotEmit=false]
     * @private
     */

  }, {
    key: 'setData',
    value: function setData(nodes) {
      var doNotEmit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

      var oldNodesData = this.body.data.nodes;

      if (nodes instanceof DataSet || nodes instanceof DataView) {
        this.body.data.nodes = nodes;
      } else if (Array.isArray(nodes)) {
        this.body.data.nodes = new DataSet();
        this.body.data.nodes.add(nodes);
      } else if (!nodes) {
        this.body.data.nodes = new DataSet();
      } else {
        throw new TypeError('Array or DataSet expected');
      }`

Despite I used a gwt.visualization DataView in my code! Is there a Java data type that I can use to pass data to your script or I need to implement mine? It will work at all?

Please advise...

**Comment by zbeedatm** [comment]
Other approach that I'm trying is:

`String strNodes = new StringBuilder("[")
			.append("{id: 1, label: 'Node 1'},")
			.append("{id: 2, label: 'Node 2'},")
			.append("{id: 3, label: 'Node 3'},")
			.append("{id: 4, label: 'Node 4'},")
			.append("{id: 5, label: 'Node 5'}")
			.append("]").toString();
		
		String strEdges = new StringBuilder("[")
			.append("{from: 1, to: 3},")
			.append("{from: 1, to: 2},")
			.append("{from: 2, to: 4},")
			.append("{from: 2, to: 5},")
			.append("{from: 3, to: 3}")
			.append("]").toString();`


and using JSNI to eval data:
`private native void loadNetwork(String nodes, String edges)/*-{  
		

	  // create a network
	  var container = $doc.getElementById('mynetwork');
	  var data = {
	    nodes: nodes,
	    edges: edges
	  };
	  var options = {};
	  
	  //var network = new $wnd.vis.Network(container, data, options);
	  var network = new $wnd.vis.Network(container, eval(data), options);
	  
		$wnd.draw();
	}-*/;`


using eval function of javascript to do the convert to DataSet/DataView:
var network = new $wnd.vis.Network(container, **eval**(data), options);

**Comment by yotamberk** [comment]
So can this be closed?

**Comment by zbeedatm** [comment]
If there are no plans to build a java wrapper jar in the future (to support GWT and other java frameworks), I believe that yes we can close it.
What about your JavaScript, is it ensured that it will run on all kind of browsers? Or need to do some customizations?

**Comment by brichbe** [comment]
@zbeedatm first I must say, it's very difficult to understand all the comments you've posted above, so it's unclear what exact issues you're having.  If you still have questions about integrating VisJs with GWT, you need to post a clear, concise comment that explains exactly what you've tried and what isn't working.  The way you're posting stack traces and small, incomplete snippets of your code is not helpful.

From what I can gather, I believe you were confusing a Java GWT class named "DataView" and you were expecting that to be automatically compatible with the VisJs [DataView](http://visjs.org/docs/data/dataview.html) simply because they share the same name?  If that's true, you must understand those two things are completely different and they are not compatible in the way you're trying to convert them.

Until you get more comfortable with using GWT, I recommend you consider integrating with VisJs in the following way:  Construct your nodes, edges, and options within your GWT Java code, each as a JSONObject with all the key-value pairs each one needs.  For each of those objects, get its String representation by calling `toString()`.  Create a JSNI method that takes each of those strings as parameters.  Have this JSNI method call a function you've declared in another javascript file within your project, where this function also accepts all those strings as parameters.  Within that js function, convert each of those strings to JavaScript objects by using `JSON.parse(<your string>)`.  Then initialize your network and your node/edge Datasets using those JSON objects.

To address your most recent comment:

I'm not aware of any existing GWT wrapper for VisJs, but you're free to implement this yourself if you find it necessary.  The maintainers of this VisJs project have many other higher priorities, so don't expect them to find it necessary to create a GWT wrapper of their own javascript lib.

Of course the goal of any javascript lib is to be compatible with all modern browsers, but issues are discovered from time to time.  If you find a browser-specific issue in the future, and you're sure the issue is not because of GWT but it's because of this VisJs lib, then post a new issue on this github project to explain what the problem is.

**Comment by zbeedatm** [comment]
The solution you are suggesting is what I did at the end and it working fine. 
My only concern was compatibility with all browsers, that is gwt bonus actually.. You can keep your priorities,  no one asked to change it,  i just said if you are planing in the future to do so. As I know VisJs is the successor of chap-links (https://almende.github.io/chap-links-library/) and they had their wrappers jars. 
I didnt use the DataView of java just because it has the same name!!! I searched for a one that was used in visualization project and it was taken from gwt-visualization jar (I think it was used in chap-links as well).  This is how I thought it should work but I realized very quick that its not and I switched my approach. 

Closing this issue.
